architecture.adoc

```diff
@@ -0,0 +1,134 @@
+
+Classes that represent concepts calculated from the <<collected-data, Collected Data>>.
+
+ApplicationNfrViolation:: A reference to an <<non-functional-requirement, NFR>> that has not been met for a given
+application with a description of why it was not met and what can be done to fix it.
+AuroraStatus:: The representation of the application status based on data collected from OpenShift and from the
+application itself from its Management Interface (ref needed).
+
+
+[[cache]]
+== Cache
+
+Used for generating API responses based on the ApplicationData. The cache contains an ApplicationData entry for each
+application running on the cluster and is populated by the Crawler (described next).
+
+
+== Crawler
+
+The core workflow for collecting Application Data from the Aurora OpenShift cluster and updating the Cache with the
+collected data.
+
+The Crawler starts by collecting a handle for each application running on the cluster, and then collects information for
+those applications in parallel. The collected information per application is called Application Data and the Cache
+is updated individually for each application as the collection process progresses.
+
+Upon completion of a crawl of the cluster, the Cache is updated to remove entries for applications that no longer exist.
+
+=== Implementation Details
+ApplicationDataCrawler:: Responsible for the core parallelized data collection workflow. Does not collect any data
+by itself and does not know of any of the underlying details of OpenShift or OpenShift objects. Handles parallelization
+and *triggering* of data collection and cache update. Workflow component.
+
+ApplicationDataService:: Responsible for collecting <<application-data, Application Data>> given an <<ApplicationId>>.
+Knows of OpenShift objects, but not how to collect them (done via OpenShiftService). Also knows how to identify which
+applications are currently running on the cluster.
+
+AuroraStatusCalculator:: Service object for determining the AuroraStatus for an application. Configured with different
+kinds of thresholds that affect status transitions (AVERAGE_RESTART_OBSERVE_THRESHOLD, AVERAGE_RESTART_ERROR_THRESHOLD,
+DIFFERENT_DEPLOYMENT_HOUR_THRESHOLD).
+
+NfrAuditor:: Service object for determining if an application violates any of the current non functional requirements.
+Operates from a set of NfrRules.
+
+NfrRule:: Implementation of a check that will generate an NfrViolation if the <<non-functional-requirement, Nrf>>
+associated with the current NfrRule has been violated for a given Application Data instance.
+
+ApplicationManagementInterface:: Accessor object for the Management Interface for a given application. *Note:* This
+may quite possibly be better implemented as a Service.
+
+OpenShiftService:: Low level communication and serialization handling.
+
+
+=== Implementation Details
+ApplicationDataCacheService:: Responsible for containing the cache of the individual ApplicationData entries.
```

Comment:
This should be implemented as a Redis DataStore. Not as in memory map. A store that can createOrUpdate an entry and delete all entries older then a given time.